### PR TITLE
STYLE: Avoid subtraction in integer inequalities

### PR DESCRIPTION
**Pull request by Leengit**
Expressions such as 0U - 1U and 3U - 5U are not negative numbers but very large positive numbers and thus inequalities such as `i < size - 5` may give counterintuitive results.  Using a regular expression substitution we have the subtracted term from the right-hand side instead added to the left-hand side:
```bash
  cd ~/git/ITK
  egrep -lR '([A-Za-z_]+) ([<>]=?) ([A-Za-z]+) - ([A-Za-z_0-9]+);' |
    fgrep -v ThirdParty |
    xargs sed -i -r -e 's/([A-Za-z_]+) ([<>]=?) ([A-Za-z]+) - ([A-Za-z_0-9]+);/\1 + \4 \2 \3;/g'
```

Pros:
1. If previously the right-hand side should have been interpreted as a negative number but was instead interpreted as a large positive number then this will fix that bug.  Although likely such bugs have already been detected, and bypassed such as via checks prior to the offending inequality, this change will future proof the code in case modifications to the code allow such cases to fall through to the inequality.

Cons:
1. Code such as `i + 1 < size` is less intuitive / understandable to my eyes than `i < size - 1`.
2. It is possible that in the previous code the compiler computes `size - 1` only once in loops that do not modify `size`.  However in the new code the compiler may have to compute `i + 1` with every iteration.

Marking this as draft until we decide whether the pros outweigh the cons.

## PR Checklist
- [X] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [X] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [X] Added test (or behavior not changed)
- [X] Updated API documentation (or API not changed)
- [X] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [X] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [X] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

### CODE REVIEW

**Comment by hjmjohnson** [comment]
@Leengit  Thank you for this thought-provoking submission.  I have some concerns over con#2 mostly.  It might have negative performance implications.

In many cases The math involves compile time constants, so a static_assert, and a constexpr would be a very very safe route.

```c++
for(int i =0; i < ImageDimenstion -1; i++)

//bcomes
static_assert(ImageDimension > 0, "Only image dimesions greate than 0 are allowed");
constexpr unsigned int OneLessImageDimension = ImageDimension - 1;
for(unsigned int i =0 ; i < OneLessImageDimension; i++)

```

File: Modules/Core/Common/include/itkConicShellInteriorExteriorSpatialFunction.hxx [hunk 1]
```diff
@@ -94,14 +94,14 @@ ConicShellInteriorExteriorSpatialFunction<VDimension, TInput>::PrintSelf(std::os
 
   unsigned int i;
   os << indent << "Origin: [";
-  for (i = 0; i < VDimension - 1; ++i)
+  for (i = 0; i + 1 < VDimension; ++i)
```
Comment (dzenanz) [inline Modules/Core/Common/include/itkConicShellInteriorExteriorSpatialFunction.hxx:97]: Do we allow Dimension to be zero? If we don't, this is not needed.

File: Modules/Core/Common/include/itkSphereSpatialFunction.hxx [hunk 2]
```diff
@@ -60,7 +60,7 @@ SphereSpatialFunction<VImageDimension, TInput>::PrintSelf(std::ostream & os, Ind
 
   unsigned int i;
   os << indent << "Center: [";
-  for (i = 0; i < VImageDimension - 1; ++i)
+  for (i = 0; i + 1 < VImageDimension; ++i)
```
Comment (dzenanz) [inline Modules/Core/Common/include/itkSphereSpatialFunction.hxx:63]: We disallow Image's dimension to be zero.
```cpp
#include "itkImage.h"
int
main()
{
  auto zeroImage = itk::Image<short, 0>::New();
}
```
gives the following compile errors:
```text
1>------ Build started: Project: tester, Configuration: Debug x64 ------
1>tester.cpp
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkIndex.h(285,28): error C2338: Error: Only positive value sized VDimension allowed
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImageBase.h(133): message : see reference to class template instantiation 'itk::Index<0>' being compiled
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(89): message : see reference to class template instantiation 'itk::ImageBase<0>' being compiled
1>C:\Misc\Tester\tester.cpp(6): message : see reference to class template instantiation 'itk::Image<short,0>' being compiled
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkOffset.h(238,28): error C2338: Error: Only positive value sized VDimension allowed
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImageBase.h(138): message : see reference to class template instantiation 'itk::Offset<0>' being compiled
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkSize.h(226,28): error C2338: Error: Only positive value sized VDimension allowed
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImageBase.h(142): message : see reference to class template instantiation 'itk::Size<0>' being compiled
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkVector.h(63,1): error C2503: 'itk::FixedArray<T,0>': base classes cannot contain zero-sized arrays
1>        with
1>        [
1>            T=itk::ImageBase<0>::SpacingValueType
1>        ]
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImageBase.h(813): message : see reference to class template instantiation 'itk::Vector<itk::ImageBase<0>::SpacingValueType,0>' being compiled
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkPoint.h(54,1): error C2503: 'itk::FixedArray<T,0>': base classes cannot contain zero-sized arrays
1>        with
1>        [
1>            T=itk::ImageBase<0>::SpacingValueType
1>        ]
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImageBase.h(814): message : see reference to class template instantiation 'itk::Point<itk::ImageBase<0>::PointValueType,0>' being compiled
1>C:\Dev\ITK-git\Modules\ThirdParty\VNL\src\vxl\core\vnl\vnl_matrix_fixed.h(109,1): error C2087: 'data_': missing subscript
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkMatrix.h(297): message : see reference to class template instantiation 'vnl_matrix_fixed<T,0,0>' being compiled
1>        with
1>        [
1>            T=itk::SpacePrecisionType
1>        ]
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImageBase.h(815): message : see reference to class template instantiation 'itk::Matrix<itk::SpacePrecisionType,0,0>' being compiled
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImageBase.h(816,17): error C2229: class 'itk::ImageBase<0>' has an illegal zero-sized array
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImageBase.h(820,17): error C2229: class 'itk::ImageBase<0>' has an illegal zero-sized array
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImageBase.h(821,17): error C2229: class 'itk::ImageBase<0>' has an illegal zero-sized array
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImageBase.h(873,19): error C2229: class 'itk::ImageBase<0>' has an illegal zero-sized array
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImageRegion.h(365,13): error C2229: class 'itk::ImageRegion<0>' has an illegal zero-sized array
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImageBase.h(875): message : see reference to class template instantiation 'itk::ImageRegion<0>' being compiled
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImageBase.h(876,14): error C2229: class 'itk::ImageBase<0>' has an illegal zero-sized array
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImageBase.h(877,14): error C2229: class 'itk::ImageBase<0>' has an illegal zero-sized array
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(89,1): error C2503: 'itk::ImageBase<0>': base classes cannot contain zero-sized arrays
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(131,3): error C3210: 'ImageBase<0>': a member using-declaration can only be applied to a base class member
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(134,3): error C3210: 'ImageBase<0>': a member using-declaration can only be applied to a base class member
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(135,3): error C3210: 'ImageBase<0>': a member using-declaration can only be applied to a base class member
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(138,3): error C3210: 'ImageBase<0>': a member using-declaration can only be applied to a base class member
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(141,3): error C3210: 'ImageBase<0>': a member using-declaration can only be applied to a base class member
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(142,3): error C3210: 'ImageBase<0>': a member using-declaration can only be applied to a base class member
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(148,3): error C3210: 'ImageBase<0>': a member using-declaration can only be applied to a base class member
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(152,3): error C3210: 'ImageBase<0>': a member using-declaration can only be applied to a base class member
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(156,3): error C3210: 'ImageBase<0>': a member using-declaration can only be applied to a base class member
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(157,3): error C3210: 'ImageBase<0>': a member using-declaration can only be applied to a base class member
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(161,3): error C3210: 'ImageBase<0>': a member using-declaration can only be applied to a base class member
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(168,3): error C3210: 'ImageBase<0>': a member using-declaration can only be applied to a base class member
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(392,3): error C3210: 'ImageBase<0>': a member using-declaration can only be applied to a base class member
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(101,3): error C3668: 'itk::Image<short,0>::CreateAnother': method with override specifier 'override' did not override any base class methods
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(104,3): error C3668: 'itk::Image<short,0>::GetNameOfClass': method with override specifier 'override' did not override any base class methods
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(190,3): error C3668: 'itk::Image<short,0>::Allocate': method with override specifier 'override' did not override any base class methods
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(195,3): error C3668: 'itk::Image<short,0>::Initialize': method with override specifier 'override' did not override any base class methods
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(381,3): error C3668: 'itk::Image<short,0>::Graft': method with override specifier 'override' did not override any base class methods
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(321,3): error C3668: 'itk::Image<short,0>::GetNumberOfComponentsPerPixel': method with override specifier 'override' did not override any base class methods
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(379,3): error C3668: 'itk::Image<short,0>::PrintSelf': method with override specifier 'override' did not override any base class methods
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(383,4): error C3668: 'itk::Image<short,0>::~Image': method with override specifier 'override' did not override any base class methods
1>C:\Dev\ITK-git\Modules\Core\Common\include\itkImage.h(391,3): error C3668: 'itk::Image<short,0>::ComputeIndexToPhysicalPointMatrices': method with override specifier 'override' did not override any base class methods
1>Done building project "tester.vcxproj" -- FAILED.
```

**Review by dzenanz — CHANGES_REQUESTED** [review]
Great majority of these changes are unnecessary. Rewriting to keep unsigned math correct makes the conditions look unintuitive. It is better to add an assert before the subset of conditions which do not involve image dimension being zero and don't have a regular check not far above (e.g. checking whether `argc<5` before using `argc - 5`).

File: Modules/Core/Common/include/itkConicShellInteriorExteriorSpatialFunction.hxx (hunk 1 above)
Comment (Leengit) [reply Modules/Core/Common/include/itkConicShellInteriorExteriorSpatialFunction.hxx:97]: Yes, I think that _**all**_ of the changes in this pull request are not needed!  I mean that in the senses that either
1. the variable's range precludes the underflow, or
2. the loop is only entered after a runtime check that the variable is in a range that will not underflow.

The added value of this pull request comes about only as insurance: if we later allow `VDimension` to be zero or, in other scenarios, if we later remove the runtime range check.

I am now convinced that this isn't enough added value ... and will close this pull request.

File: Modules/Core/Common/include/itkSphereSpatialFunction.hxx (hunk 2 above)
Comment (Leengit) [reply Modules/Core/Common/include/itkSphereSpatialFunction.hxx:63]: Thank you for verifying that.

**Comment by Leengit** [comment]
I am convinced that the cons outweigh the pros.  I am closing this pull request.

**Pull request closed without merging**